### ISSUE TITLE
Question: Any chance to restore in an unprivileged container?

### ISSUE THREAD

**Issue by wata727** [checkpoint-restore/criu#2322]
**Description**

I'm looking into CRIU as an option to speed up a slow-starting process that is run as a container. After looking into some issues, I understand that `--privileged` is required to invoke `criu restore` inside a container, but this constraint cannot apply to widely used serverless solutions such as [AWS Fargate](https://aws.amazon.com/fargate/), [Google Cloud Run](https://cloud.google.com/run), etc.

If you are restoring a trivial process that does not require restoring PID or network configurations, is it possible to restore it inside an unprivileged container?

As far as I investigated, it seemed impossible to overcome this constraint because CAP_SYS_RESOURCE is ultimately required for prctl(2) + PR_SET_MM below, but is this understanding correct?
https://github.com/checkpoint-restore/criu/blob/v3.19/criu/pie/restorer.c#L1922

If this understanding is correct, are there any plans to change CRIU or the Linux kernel in the future to allow restores from inside an unprivileged container?

**Comment by adrianreber** [comment]
To run CRIU as non root we introduced CAP_CHECKPOINT_RESTORE. That is the minimum you need for really simple processes.

**Comment by wata727** [comment]
Thanks @rst0git, I understand that additional capabilities are required for complete restoring, but I wanted to know if there was a way to restore a trivial process (e.g. [simple loop](https://criu.org/Simple_loop)) in an unprivileged container.

My use case only requires processing that is not directly related to the kernel, such as file loading, parsing, and loading Ruby classes, so if this is possible, it would be possible to speed up the startup process while still using Docker.

After digging deeper into the cause of the restore failure, I found that it was failing with capset(2).
https://github.com/checkpoint-restore/criu/blob/50aa6da65adbeea50152753b86fb02f22bb88a22/criu/pie/restorer.c#L341

The dumped process was running in a privileged container, so trying to set all the capabilities here was causing an error. I restored a process running in an unprivileged container with `criu dump -t <pid> --unprivileged -j` and everything worked perfectly fine.

My conclusion here is that vendors like AWS and Google may be able to support `CAP_CHECKPOINT_RESTORE` in containers in the future to achieve faster startups using CRIU. Thank you both for answering my questions.
